functions/_common/analytics.ts

```diff
@@ -0,0 +1,133 @@
+
+    try {
+        const response = await fetch(ga4Endpoint, {
+            method: "POST",
+            headers: {
+                "Content-Type": "application/json",
+            },
+            body: JSON.stringify({
+                client_id,
+                events: [event],
+            }),
+        })
+        if (!response.ok) {
+            console.error(
+                `GA4 MP request failed with status: ${response.status}`,
+                await response.text()
+            )
+        } else {
+            console.log("GA4 MP event sent successfully.")
+        }
+    } catch (error) {
+        console.error("Error sending event to GA4 MP:", error)
+    }
+}
+
+export function getCommonEventParams(request: Request, pathname: string) {
+    const url = new URL(request.url)
+    return {
+        page_location: url.href,
+        page_path: pathname,
+        referrer: request.headers.get("referer") || "(direct)",
+        user_agent: request.headers.get("user-agent") || "",
+        method: request.method,
+        country: request.headers.get("cf-ipcountry") || "unknown",
+    }
+}
+
+function validateAnalyticsEnvVariables(env: Env): boolean {
+    return (
+        !!env.CLOUDFLARE_GOOGLE_ANALYTICS_MEASUREMENT_PROTOCOL_KEY &&
+        !!env.CLOUDFLARE_GOOGLE_ANALYTICS_MEASUREMENT_ID &&
+        !!env.CLOUDFLARE_GOOGLE_ANALYTICS_SAMPLING_RATE &&
+        !!env.CLOUDFLARE_GOOGLE_ANALYTICS_SESSION_COOKIE_NAME
+    )
+}
+
+export const analyticsMiddleware: PagesFunction<Env> = async (context) => {
+    const { request, env } = context
+    if (!validateAnalyticsEnvVariables(env)) {
+        return context.next()
+    }
+    const url = new URL(request.url)
+    const pathname = url.pathname
+
+    // Check if this is a request we want to track
+    const downloadFormat = getDownloadFormat(pathname)
+
+    if (downloadFormat && shouldSample(env)) {
+        // Get the response first to capture status code
+        const response = await context.next()
+
+        // Use waitUntil to send analytics without blocking the response
+        context.waitUntil(
+            sendEventToGA4(
+                {
+                    name: "cf_download",
+                    params: {
+                        ...getCommonEventParams(request, pathname),
+                        format: downloadFormat,
+                        status_code: response.status,
+                    },
+                },
+                request,
+                env
+            )
+        )
+
+        return response
+    }
+
+    return context.next()
+}
+
+function shouldSample(env: Env): boolean {
+    return (
+        Math.random() < parseInt(env.CLOUDFLARE_GOOGLE_ANALYTICS_SAMPLING_RATE)
+    )
+}
+
+// Helper function to detect download format
+function getDownloadFormat(pathname: string): string | null {
```

Comment:
I'm not sure if we want to treat our json configs as data download. It should be mostly us, who fetch them. Have you confirmed this with Daniel/Bobbie? They are not included in the bundle when you download data from the data download. But if we want them, the we should probably tread images as "downloads" too.

And more generally, I thought we'll be logging every request (when sampled), not just downloads. Did I misunderstand?

Reply:
After chatting with Daniel, we're just gonna run analytics sampling on every route, not just downloads (so I've refactored the commit to be more agnostic)